Add imx95 SoC and EVK board support

Zhiqiang-Hou opened the pull request:
Reworked the i.MX9 series SoC support: made the `imx9` directory as the SoC series hierarchy and moved the i.MX93 support code to `imx9/imx93`.
And then added i.MX95 SoC and EVK board support (including A55 and M7 clusters).

Dependency:
- https://github.com/zephyrproject-rtos/hal_nxp/pull/406

Zhiqiang-Hou committed:
dts: arm64: nxp: add device tree for i.MX95 Cortex-A55

Add DTSi for i.MX95 Cortex-A55.

Signed-off-by: Hou Zhiqiang <Zhiqiang.Hou@nxp.com>
Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

Zhiqiang-Hou commented:
depends on https://github.com/zephyrproject-rtos/hal_nxp/pull/406

Zhiqiang-Hou commented:
Update: Fixed compliance check errors and tests with twister issue.

JiafeiPan reviewed (commented):
As imx93 and imx95 have explicit difference because of SM, so they can't share most of the code, I suggest to change to use the following SoC directory layout:
```
soc/nxp/imx/imx9
├── imx93
│   └── a55
└── imx95
    ├── a55
    └── m7
```

LaurentiuM1234 commented:
would it be possible to use https://github.com/zephyrproject-rtos/zephyr/pull/75102 for the SCMI support instead of relying on the HAL support for it?

dbaluta commented:
@Zhiqiang-Hou for people following git history and trying to understand how a support for a cpu is added do you think it would be possible to separate patches to for different cores? 

I think it is fine to add everything in one single PR, but we should somehow better reformat patches and add more documenation and explanations in the commit messages.

yangbolu1991 committed:
dts: arm64: nxp: add device tree for i.MX95 Cortex-M7

Added device tree file for i.MX95 Cortex-M7 and added i.MX95
clock ID header file.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>
Signed-off-by: Hou Zhiqiang <Zhiqiang.Hou@nxp.com>
Signed-off-by: Laurentiu Mihalcea <laurentiu.mihalcea@nxp.com>

yangbolu1991 committed:
soc: nxp: imx: add i.MX95 Cortex-M7 support

The i.MX95 applications processor features advanced graphics and
video cores, powerful vision and machine learning acceleration,
efficient CPU performance, real-time processing, and advanced
security with the integrated EdgeLock® secure enclave to support
energy-efficient edge computing.

Signed-off-by: Hou Zhiqiang <Zhiqiang.Hou@nxp.com>
Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>
Signed-off-by: Laurentiu Mihalcea <laurentiu.mihalcea@nxp.com>

yangbolu1991 committed:
boards: nxp: add imx95_evk M7 support

The IMX95LPD5EVK-19 board is a platform designed to show the most
commonly used features of the i.MX 95 automotive applications
processor. It is an entry-level development board, which helps
developers to get familiar with the processor before investing a
large amount of resources in more specific designs.

Signed-off-by: Hou Zhiqiang <Zhiqiang.Hou@nxp.com>
Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>
Signed-off-by: Laurentiu Mihalcea <laurentiu.mihalcea@nxp.com>

yangbolu1991 committed:
driver: pinctrl: add SCMI-based pinctrl driver for NXP i.MX SoCs

On some i.MX SoCs, such as i.MX95, the System Manager is running on a
Cortex-M core to manage the hardware resources and provide services for
SCMI requests.

So add the SCMI-based pinctrl driver to support these i.MX SoCs.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>
Signed-off-by: Hou Zhiqiang <Zhiqiang.Hou@nxp.com>
Signed-off-by: Laurentiu Mihalcea <laurentiu.mihalcea@nxp.com>

Zhiqiang-Hou committed:
soc: nxp: imx: add i.MX95 Cortex-A55 support

Added basic soc support for i.MX95 Cortex-A55.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>
Signed-off-by: Hou Zhiqiang <Zhiqiang.Hou@nxp.com>

Zhiqiang-Hou committed:
boards: nxp: add imx95_evk A55 support

Added basic board support for imx95_evk A55.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>
Signed-off-by: Hou Zhiqiang <Zhiqiang.Hou@nxp.com>

yangbolu1991 committed:
drivers: mbox: make mbox_nxp_imx_mu use PRE_KERNEL_1 init level

The MU init should be very early for some platforms. For examples,
i.MX95 is using MU for ARM SCMI. It should be initialzied early
for next power, pinctrl, clock etc. management.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

dbaluta commented:
@yangbolu1991  PR looks good to me. Please check if the compliance check can be fixed and we are good to go.

LaurentiuM1234 commented on soc/nxp/imx/imx95/Kconfig.defconfig.mimx95.m7:
```diff
@@ -0,0 +1,20 @@
+if SOC_MIMX9596_M7
+
+config FLASH_SIZE
+	default $(dt_node_reg_size_int,/soc/itcm@0,0,K)
```
nitpick: we can use `zephyr,flash` here which references `itcm`. Same applies to `FLASH_BASE_ADDRESS`.

yangbolu1991 replied:
Okay.

LaurentiuM1234 commented on boards/nxp/imx95_evk/imx95_evk-pinctrl.dtsi:
```diff
@@ -0,0 +1,29 @@
+			pinmux = <&iomuxc_uart1_rxd_lpuart_rx_lpuart1_rx>,
+				<&iomuxc_uart1_txd_lpuart_tx_lpuart1_tx>;
+			bias-pull-up;
+			slew-rate = "slightly_fast";
```
What's the reason for using different slew rate/drive strength values for this UART (compared to `lpuart3`)?

yangbolu1991 replied:
Oh, that's added by me and Zhiqiang Seperately.
Let me change them all to default slightly_fast and x4 since no specific requirement.
Thanks.

yangbolu1991 replied:
Checked. The configuration is same with linux dts.

LaurentiuM1234 commented on dts/arm/nxp/nxp_imx95_m7.dtsi:
```diff
@@ -0,0 +1,154 @@
+	};
+
+	firmware {
+		scmi_dev: scmi {
```
Do we need to reference this node somewhere? If not, then the label shouldn't be needed.

yangbolu1991 replied:
Okay. It can be removed.

LaurentiuM1234 commented on dts/arm64/nxp/nxp_mimx95_a55.dtsi:
```diff
@@ -0,0 +1,242 @@
+	};
+
+	firmware {
+		scmi_dev: scmi {
```
Same as M7 DTS.

LaurentiuM1234 commented on soc/nxp/imx/imx95/a55/mmu_regions.c:
```diff
@@ -0,0 +1,77 @@
+	MMU_REGION_DT_COMPAT_FOREACH_FLAT_ENTRY(nxp_kinetis_lpuart,
+			 (MT_DEVICE_nGnRnE | MT_P_RW_U_NA | MT_NS))
+
+#if CONFIG_SOF
```
We can remove these. They're not defined anywhere in the DTS anyways.

yangbolu1991 replied:
Okay.

LaurentiuM1234 reviewed (commented):
Looks good! A couple of comments below

LaurentiuM1234 commented:
> @yangbolu1991 PR looks good to me. Please check if the compliance check can be fixed and we are good to go.

This PR depends on https://github.com/zephyrproject-rtos/zephyr/pull/75102 so the compliance check failures are to be expected.

yangbolu1991 commented:
Update the pull request for v3 version. Changes,
- Improved code per @LaurentiuM1234 suggestion.
- Fixed one compliance check.
- dependency: 
    - SCMI: #75102 
    - i.MX95 HAL: https://github.com/zephyrproject-rtos/hal_nxp/pull/406

Thank you very much.

Zhiqiang-Hou commented:
> As imx93 and imx95 have explicit difference because of SM, so they can't share most of the code, I suggest to change to use the following SoC directory layout:
> 
> ```
> soc/nxp/imx/imx9
> ├── imx93
> │   └── a55
> └── imx95
>     ├── a55
>     └── m7
> ```

Thanks a lot for the suggestion, @yangbolu1991 has add a separate directory for imx95 under soc/nxp/imx, I think it also make sense.

Zhiqiang-Hou commented:
> @Zhiqiang-Hou for people following git history and trying to understand how a support for a cpu is added do you think it would be possible to separate patches for different cores?
> 
> I think it is fine to add everything in one single PR, but we should somehow better reformat patches and add more documenation and explanations in the commit messages.

Thanks a lot for the suggestions! Actually we separated A-core and M-core support code in different patches, and just submitted the code in a single pull-request.

Zhiqiang-Hou commented on soc/nxp/imx/imx95/a55/mmu_regions.c:
```diff
@@ -0,0 +1,51 @@
+			      DT_REG_ADDR(DT_NODELABEL(iomuxc)),
+			      DT_REG_SIZE(DT_NODELABEL(iomuxc)),
+			      MT_DEVICE_nGnRnE | MT_P_RW_U_NA | MT_NS),
+#endif
```
Remove this block as we have separated imx95 code into a new directory.

yangbolu1991 replied:
Okay.

Zhiqiang-Hou commented:
> @JiafeiPan @dbaluta @LaurentiuM1234 I updated the pull request. The changes are,
> 
> * Used ARM SCMI [[RFC] Introduce support for ARM's SCMI #75102](https://github.com/zephyrproject-rtos/zephyr/pull/75102), instead of hal imx-sm.
> * Since i.MX95 was based on SCMI pretty different from i.MX93 in software, moved it into new soc folder.
> 
> Thanks.

Zhiqiang-Hou commented on boards/nxp/imx95_evk/doc/index.rst:
```diff
@@ -0,0 +1,121 @@
++-----------+------------+-------------------------------------+
+| UART      | on-chip    | serial port                         |
++-----------+------------+-------------------------------------+
+
```
The above table is for A-core target, add a new table to list the hardware features used by M-core.

yangbolu1991 replied:
Okay.

Zhiqiang-Hou commented on boards/nxp/imx95_evk/doc/index.rst:
```diff
@@ -0,0 +1,121 @@
+
+This board configuration uses a system clock frequency of 24 MHz.
+Cortex-A55 Core runs up to 1.8 GHz.
+
```
Append M-core details?

yangbolu1991 replied:
Okay.

Zhiqiang-Hou commented:
Sorry, I'd know how I closed the PR, reopen it.

Zhiqiang-Hou commented on boards/nxp/imx95_evk/doc/index.rst:
```diff
@@ -0,0 +1,121 @@
+    thread_a: Hello World from cpu 0 on imx95_evk!
+    thread_b: Hello World from cpu 0 on imx95_evk!
+    thread_a: Hello World from cpu 0 on imx95_evk!
+
```
Yangbo, add quick guide for M-core target.

yangbolu1991 replied:
Okay.

Zhiqiang-Hou commented on drivers/pinctrl/Kconfig.imx:
```diff
@@ -3,7 +3,7 @@
 	bool "Pin controller driver for iMX MCUs"
-	depends on DT_HAS_NXP_IMX_IOMUXC_ENABLED
+	depends on DT_HAS_NXP_IMX_IOMUXC_ENABLED || ARM_SCMI_PINCTRL_HELPERS
 	depends on HAS_MCUX_IOMUXC || HAS_IMX_IOMUXC
```
still depends on the IOMUXC HAL driver when switched to the SCMI one?

Zhiqiang-Hou commented on drivers/pinctrl/pinctrl_imx.c:
```diff
@@ -8,6 +8,76 @@
+}
+
+#else
+
```
Move to a separate SCMI-based driver, like we did when based on imx-sm helpers?

Zhiqiang-Hou commented on drivers/pinctrl/pinctrl_imx.c:
```diff
@@ -79,6 +149,8 @@ int pinctrl_configure_pins(const pinctrl_soc_pin_t *pins, uint8_t pin_cnt,
 
+#endif /* CONFIG_ARM_SCMI_PINCTRL_HELPERS */
+
 static int imx_pinctrl_init(void)
```
bracket the calling of HAL clock helpers by the condition !CONFIG_ARM_SCMI_PINCTRL_HELPERS.
But I prefer to create a new driver for SCMI-based one and leave this driver as it's.

yangbolu1991 replied:
For me, either keeping one driver or adding a new driver is okay.
I will add a new driver to address your above comments.
Thanks.

Zhiqiang-Hou commented on soc/nxp/imx/imx95/Kconfig:
```diff
@@ -0,0 +1,25 @@
+	select CPU_HAS_ARM_MPU
+	select ARM_MPU
+	select HAS_MCUX
+	select HAS_MCUX_IOMUXC
```
remove selecting of MCUX IOMUXC?

Zhiqiang-Hou commented on soc/nxp/imx/imx95/Kconfig:
```diff
@@ -0,0 +1,25 @@
+	select CPU_CORTEX_A55
+	select ARM_ARCH_TIMER if SYS_CLOCK_EXISTS
+	select HAS_MCUX
+	select HAS_MCUX_IOMUXC
```
remove selecting of MCUX IOMUXC?

Zhiqiang-Hou commented on soc/nxp/imx/imx95/a55/CMakeLists.txt:
```diff
@@ -0,0 +1,5 @@
+
+zephyr_include_directories(.)
+zephyr_sources_ifdef(CONFIG_ARM_MMU mmu_regions.c)
+set(SOC_LINKER_SCRIPT ${CMAKE_CURRENT_SOURCE_DIR}/linker.ld CACHE INTERNAL "")
```
Use the ARM64 common linker as NO dependence on extra sections for SOF:
`set(SOC_LINKER_SCRIPT ${ZEPHYR_BASE}/include/zephyr/arch/arm64/scripts/linker.ld CACHE INTERNAL "")`

Zhiqiang-Hou commented on soc/nxp/imx/imx95/a55/linker.ld:
```diff
@@ -0,0 +1,366 @@
+     * calculate this value here. */
+    _flash_used = LOADADDR(.last_section) + SIZEOF(.last_section) - __rom_region_start;
+
+}
```
Remove this linker and change to use the ARM64 common linker.

yangbolu1991 commented:
Update the pull request for v4 version to address @Zhiqiang-Hou 's comments. Changes,

- Moved scmi pinctrl code to a new pinctrl_imx_scmi driver. Removed dependency on HAS_MCUX_IOMUXC.
- Converted to use ARM64 common linker file for A55.
- Removed unused mmu region code for A55.
- Reworked board doc and added guide for M7.

dependency:
- SCMI: https://github.com/zephyrproject-rtos/zephyr/pull/75102
- i.MX95 HAL: https://github.com/zephyrproject-rtos/hal_nxp/pull/406

Thank you very much.

JiafeiPan commented:
> > As imx93 and imx95 have explicit difference because of SM, so they can't share most of the code, I suggest to change to use the following SoC directory layout:
> > ```
> > soc/nxp/imx/imx9
> > ├── imx93
> > │   └── a55
> > └── imx95
> >     ├── a55
> >     └── m7
> > ```
> 
> Thanks a lot for the suggestion, @yangbolu1991 has add a separate directory for imx95 under soc/nxp/imx, I think it also make sense.

I still think we need to put imx93 and imx95 (and maybe include imx91 in future) into a serial of imx9, yes, firstly we need a patch to move imx93 to a separate directory in imx9.

yangbolu1991 commented:
> > > As imx93 and imx95 have explicit difference because of SM, so they can't share most of the code, I suggest to change to use the following SoC directory layout:
> > > ```
> > > soc/nxp/imx/imx9
> > > ├── imx93
> > > │   └── a55
> > > └── imx95
> > >     ├── a55
> > >     └── m7
> > > ```
> > 
> > 
> > Thanks a lot for the suggestion, @yangbolu1991 has add a separate directory for imx95 under soc/nxp/imx, I think it also make sense.
> 
> I still think we need to put imx93 and imx95 (and maybe include imx91 in future) into a serial of imx9, yes, firstly we need a patch to move imx93 to a separate directory in imx9.

In my opinion, each imx soc directory keeps only a few files with minor code.
It's not very necessary to consider code reusing with using a soc series directory.
This makes deeper directory structure, and larger Kconfig/Makefile to maintain, and handling difference in code between socs is also needed.
Actually I prefer the previous way before imx9. Keep one directory for each soc, but not for soc series.
Renaming current imx9 to imx93 under soc/nxp/imx/ is the way.

But I am not insisting. 
Thanks.

yangbolu1991 commented on boards/nxp/imx95_evk/doc/index.rst:
```diff
@@ -0,0 +1,202 @@
+| SYSTICK   | on-chip    | systick                             | M7        |
++-----------+------------+-------------------------------------+-----------+
+| UART      | on-chip    | serial port                         | A55/M7    |
++-----------+------------+-------------------------------------+-----------+
```
okay, thanks.

yangbolu1991 commented on boards/nxp/imx95_evk/board.yml:
```diff
@@ -0,0 +1,5 @@
+  name: imx95_evk
+  vendor: nxp
+  socs:
+    - name: mimx9596
```
okay.

Zhiqiang-Hou replied:
I'll add SMP variants later.

yangbolu1991 commented on soc/nxp/imx/imx95/a55/mmu_regions.c:
```diff
@@ -0,0 +1,34 @@
+/*
+ * Copyright 2022 NXP
```
Should be 2024. Will fix.
Thanks.

Zhiqiang-Hou commented on boards/nxp/imx95_evk/doc/index.rst:
```diff
@@ -0,0 +1,202 @@
+- i.MX 95 automotive applications processor
+
+  - The processor integrates up to six Arm Cortex-A55 cores, and supports
+    functional safety with built-in Arm Cortex-M33 and -M7 cores
```
s/-M7/Cortex-M7 ?

yangbolu1991 replied:
These are copied from i.MX95 EVK board user guide.
It's Arm if you check https://www.arm.com/
Thanks.

Zhiqiang-Hou replied:
It's okay to follow the Arm company. But weird, generally acronyms are made up of capital letters.

Zhiqiang-Hou replied:
It's okay to follow the Arm company. But weird, generally acronyms are made up of capital letters.

Zhiqiang-Hou commented on boards/nxp/imx95_evk/doc/index.rst:
```diff
@@ -0,0 +1,202 @@
+  - The processor integrates up to six Arm Cortex-A55 cores, and supports
+    functional safety with built-in Arm Cortex-M33 and -M7 cores
+
+- DRAM memory: 128-Gbit LPDDR5 DRAM
```
Revert to `16GB` ?

yangbolu1991 replied:
Supports 128-Gbit LPDDR5 DRAM (MT62F4G32D8DV-023 AAT:C)

yangbolu1991 commented on drivers/pinctrl/Kconfig.imx:
```diff
@@ -1,4 +1,4 @@
-# Copyright (c) 2022 NXP
+# Copyright 2022, 2024 NXP
```
There should be a blank here.

JiafeiPan commented:
> > > > As imx93 and imx95 have explicit difference because of SM, so they can't share most of the code, I suggest to change to use the following SoC directory layout:
> > > > ```
> > > > soc/nxp/imx/imx9
> > > > ├── imx93
> > > > │   └── a55
> > > > └── imx95
> > > >     ├── a55
> > > >     └── m7
> > > > ```
> > > 
> > > 
> > > Thanks a lot for the suggestion, @yangbolu1991 has add a separate directory for imx95 under soc/nxp/imx, I think it also make sense.
> > 
> > 
> > I still think we need to put imx93 and imx95 (and maybe include imx91 in future) into a serial of imx9, yes, firstly we need a patch to move imx93 to a separate directory in imx9.
> 
> In my opinion, each imx soc directory keeps only a few files with minor code. It's not very necessary to consider code reusing with using a soc series directory. This makes deeper directory structure, and larger Kconfig/Makefile to maintain, and handling difference in code between socs is also needed. Actually I prefer the previous way before imx9. Keep one directory for each soc, but not for soc series. Renaming current imx9 to imx93 under soc/nxp/imx/ is the way.
> 
> But I am not insisting. Thanks.

Hi, @yangbolu1991 , please refer to Zephyr document diagram "Hardware support Hierarchy" at https://docs.zephyrproject.org/latest/hardware/porting/board_porting.html,  For imx95, SoC family is imx, SoC serial is imx9, SoC is imx95, let's follow this hierarchy layout for imx95, thanks.

Zhiqiang-Hou committed:
boards: nxp: imx95_evk: add SMP variant support

There are 6 Cortex-A55 cores on i.MX95 EVK, this SMP variant will release
all the 6 cores to run Zephyr.

Signed-off-by: Hou Zhiqiang <Zhiqiang.Hou@nxp.com>

yangbolu1991 commented:
Updated the pull request for v5 version. Changes,

- Moved imx93/imx95 under imx9 series directory.
- Fixed copyright and doc per @JiafeiPan suggestion.

dependency:

- SCMI: https://github.com/zephyrproject-rtos/zephyr/pull/75102
- i.MX95 HAL: https://github.com/zephyrproject-rtos/hal_nxp/pull/406

If @Zhiqiang-Hou  wants to add board SMP variants, you can rebase on this.
Thank you very much.

Zhiqiang-Hou commented:
Updated the pull request for v6 version:

- Added SMP variant for A55 cluster.
- Removed the disable I/D-cache operation from the steps to run A core application.
- Corrected the indentation in board.yml.


dependency:
- SCMI: https://github.com/zephyrproject-rtos/zephyr/pull/75102
- i.MX95 HAL: https://github.com/zephyrproject-rtos/hal_nxp/pull/406

JiafeiPan commented on boards/nxp/imx95_evk/imx95_evk_mimx9596_a55_smp.dts:
```diff
@@ -0,0 +1,37 @@
+		zephyr,console = &lpuart1;
+		zephyr,shell-uart = &lpuart1;
+		zephyr,sram = &sram0;
+	};
```
SMP runs on all A55 Core? suggest SMP variants runs on two CPU Core, so that it can demo how to select SMP runs on which CPU Core.

And also update board document for introduction of how SMP variants works and how to run it.

Zhiqiang-Hou replied:
I think currently we'd better add the SMP support with all A55 cores, and use the U-Boot 'go' command to let the boot core bring up the secondary cores, as the 'cpu' command in U-Boot cmd/mp.c has been deprecated in upstream, which is used to kick the secondary cores, and the new 'cpu' command in cmd/cpu.c is still not supporting release secondary cores from reset. I'm planning to add the 'release' subcommand to the new 'cpu' command, and will update the document when it is accepted.

JiafeiPan replied:
okay, thanks.

JiafeiPan commented on soc/nxp/imx/imx9/Kconfig.defconfig:
```diff
@@ -1,8 +1,6 @@
-# Copyright 2022,2024 NXP
+# Copyright 2024 NXP
```
Don't remove 2022?

yangbolu1991 replied:
I moved files under imx9 to imx9/imx93. So, this is a actually a new file I write manually.

JiafeiPan replied:
okay, thanks.

JiafeiPan commented on soc/nxp/imx/imx9/Kconfig.soc:
```diff
@@ -1,32 +1,11 @@
-# Copyright 2022,2024 NXP
```
Don't remove 2022?

yangbolu1991 replied:
I moved files under imx9 to imx9/imx93. So, this is a actually a new file I write manually.

JiafeiPan replied:
okay, thanks.

Zhiqiang-Hou commented on soc/nxp/imx/soc.yml:
```diff
@@ -48,6 +48,12 @@ family:
       - name: a55
       - name: m33
+  - name: imx95
+    socs:
```
good catch! will remove these lines.

LaurentiuM1234 commented on dts/arm/nxp/nxp_imx95_m7.dtsi:
```diff
@@ -0,0 +1,154 @@
+		};
+	};
+
+	reserved-memory {
```
suggestion: drop this node. It has no real meaning for us. It was originally added to mirror Linux DT. I'd also suggest renaming `scmi_res0` to `scmi_shmem0` or something like that.

JiafeiPan replied:
I think we could keep to use reserve memory, so it is clear that these memory regions are reserved common memory, but not hardware module's memory address, although it has no explicit function, but make it is readable.

LaurentiuM1234 commented on dts/arm/nxp/nxp_imx95_m7.dtsi:
```diff
@@ -0,0 +1,154 @@
+		};
+	};
+
+	firmware {
```
suggestion: drop the `firmware` node. See comment from `reserved-memory` node.

yangbolu1991 replied:
I think firmware here makes it readable too as linux does.
Can we keep it too?

JiafeiPan replied:
As a linuxer, okay for me.

LaurentiuM1234 commented on dts/arm/nxp/nxp_imx95_m7.dtsi:
```diff
@@ -0,0 +1,154 @@
+			reg = <0x20000000 DT_SIZE_K(256)>;
+		};
+
+		lpuart1: serial@44380000 {
```
can we have these ordered by unit address?

yangbolu1991 replied:
Okay, thanks!

LaurentiuM1234 commented on soc/nxp/imx/imx9/imx95/m7/linker.ld:
```diff
@@ -0,0 +1,7 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/arch/arm/cortex_m/scripts/linker.ld>
```
we can just include the architecture linker script directly using CMakeLists.txt. Example: https://github.com/zephyrproject-rtos/zephyr/pull/77228.

yangbolu1991 replied:
Okay, thanks!

LaurentiuM1234 commented on dts/arm64/nxp/nxp_mimx95_a55.dtsi:
```diff
@@ -0,0 +1,242 @@
+		status = "okay";
+	};
+
+	reserved-memory {
```
same as M7 DT

LaurentiuM1234 commented on dts/arm64/nxp/nxp_mimx95_a55.dtsi:
```diff
@@ -0,0 +1,242 @@
+		};
+	};
+
+	firmware {
```
same as M7 DT

LaurentiuM1234 commented on dts/arm64/nxp/nxp_mimx95_a55.dtsi:
```diff
@@ -0,0 +1,242 @@
+		};
+	};
+
+	lpuart1: serial@44380000 {
```
same as M7 DT. can we have these sorted by unit address?

LaurentiuM1234 reviewed (commented):
SCMI dependency merged. We can proceed with merging this PR as well now. Left a couple of comments (mostly nits).

yangbolu1991 commented:
Updated to v8. Changes,
- Rebased to resolve conflict.
- dts improvement and included link script in CMakeLists.txt per @LaurentiuM1234 suggestion.

Dependency
- hal support: https://github.com/zephyrproject-rtos/hal_nxp/pull/406

Thanks a lot.

LaurentiuM1234 reviewed (approved):
IMO good to go after the HAL stuff gets merged

dbaluta commented:
@Zhiqiang-Hou 

What is with this warnings:


```

GitHub Actions / Run compliance checks on patch series (PR)

You may want to run clang-format on this change

soc/nxp/imx/imx9/imx95/pinctrl_soc.h:90 
-#define IOMUXC_PINMUX(node_id)						\
-	{									\
-	  .mux_register = DT_PROP_BY_IDX(node_id, pinmux, 0),			\
-	  .config_register = DT_PROP_BY_IDX(node_id, pinmux, 4),		\
-	  .input_register = DT_PROP_BY_IDX(node_id, pinmux, 2),			\
-	  .mux_mode = DT_PROP_BY_IDX(node_id, pinmux, 1),			\
-	  .input_daisy = DT_PROP_BY_IDX(node_id, pinmux, 3),			\
+#define IOMUXC_PINMUX(node_id)                                                                     \
+	{                                                                                          \
+		.mux_register = DT_PROP_BY_IDX(node_id, pinmux, 0),                                \
+		.config_register = DT_PROP_BY_IDX(node_id, pinmux, 4),                             \
+		.input_register = DT_PROP_BY_IDX(node_id, pinmux, 2),                              \
+		.mux_mode = DT_PROP_BY_IDX(node_id, pinmux, 1),                                    \
+		.input_daisy = DT_PROP_BY_IDX(node_id, pinmux, 3),                                 \
 	}
 
-#define Z_PINCTRL_PINMUX(group_id, pin_prop, idx)				\
+#define Z_PINCTRL_PINMUX(group_id, pin_prop, idx)                                                  \
 	IOMUXC_PINMUX(DT_PHANDLE_BY_IDX(group_id, pin_prop, idx))
 
-#define Z_PINCTRL_STATE_PIN_INIT(group_id, pin_prop, idx)			\
-	{									\
-	  .pinmux = Z_PINCTRL_PINMUX(group_id, pin_prop, idx),			\
-	  .pin_ctrl_flags = Z_PINCTRL_IOMUXC_PINCFG_INIT(group_id),		\
+#define Z_PINCTRL_STATE_PIN_INIT(group_id, pin_prop, idx)                                          \
+	{                                                                                          \
+		.pinmux = Z_PINCTRL_PINMUX(group_id, pin_prop, idx),                               \
+		.pin_ctrl_flags = Z_PINCTRL_IOMUXC_PINCFG_INIT(group_id),                          \
 	},
 
-
-#define Z_PINCTRL_STATE_PINS_INIT(node_id, prop)				\
-	{DT_FOREACH_CHILD_VARGS(DT_PHANDLE(node_id, prop),			\
-		DT_FOREACH_PROP_ELEM, pinmux, Z_PINCTRL_STATE_PIN_INIT)};	\
-
+#define Z_PINCTRL_STATE_PINS_INIT(node_id, prop)                                                   \
+	{DT_FOREACH_CHILD_VARGS(DT_PHANDLE(node_id, prop), DT_FOREACH_PROP_ELEM, pinmux,           \
+				Z_PINCTRL_STATE_PIN_INIT)};
```

Are they valid.

Also, what is why is twister failing?

```
 Run tests with twister / twister-build (1) (pull_request_target) Failing after 39m 
```

Zhiqiang-Hou commented:
> @Zhiqiang-Hou
> 
> What is with this warnings:
> 
> ```
> 
> GitHub Actions / Run compliance checks on patch series (PR)
> 
> You may want to run clang-format on this change
> 
> soc/nxp/imx/imx9/imx95/pinctrl_soc.h:90 
> -#define IOMUXC_PINMUX(node_id)						\
> -	{									\
> -	  .mux_register = DT_PROP_BY_IDX(node_id, pinmux, 0),			\
> -	  .config_register = DT_PROP_BY_IDX(node_id, pinmux, 4),		\
> -	  .input_register = DT_PROP_BY_IDX(node_id, pinmux, 2),			\
> -	  .mux_mode = DT_PROP_BY_IDX(node_id, pinmux, 1),			\
> -	  .input_daisy = DT_PROP_BY_IDX(node_id, pinmux, 3),			\
> +#define IOMUXC_PINMUX(node_id)                                                                     \
> +	{                                                                                          \
> +		.mux_register = DT_PROP_BY_IDX(node_id, pinmux, 0),                                \
> +		.config_register = DT_PROP_BY_IDX(node_id, pinmux, 4),                             \
> +		.input_register = DT_PROP_BY_IDX(node_id, pinmux, 2),                              \
> +		.mux_mode = DT_PROP_BY_IDX(node_id, pinmux, 1),                                    \
> +		.input_daisy = DT_PROP_BY_IDX(node_id, pinmux, 3),                                 \
>  	}
>  
> -#define Z_PINCTRL_PINMUX(group_id, pin_prop, idx)				\
> +#define Z_PINCTRL_PINMUX(group_id, pin_prop, idx)                                                  \
>  	IOMUXC_PINMUX(DT_PHANDLE_BY_IDX(group_id, pin_prop, idx))
>  
> -#define Z_PINCTRL_STATE_PIN_INIT(group_id, pin_prop, idx)			\
> -	{									\
> -	  .pinmux = Z_PINCTRL_PINMUX(group_id, pin_prop, idx),			\
> -	  .pin_ctrl_flags = Z_PINCTRL_IOMUXC_PINCFG_INIT(group_id),		\
> +#define Z_PINCTRL_STATE_PIN_INIT(group_id, pin_prop, idx)                                          \
> +	{                                                                                          \
> +		.pinmux = Z_PINCTRL_PINMUX(group_id, pin_prop, idx),                               \
> +		.pin_ctrl_flags = Z_PINCTRL_IOMUXC_PINCFG_INIT(group_id),                          \
>  	},
>  
> -
> -#define Z_PINCTRL_STATE_PINS_INIT(node_id, prop)				\
> -	{DT_FOREACH_CHILD_VARGS(DT_PHANDLE(node_id, prop),			\
> -		DT_FOREACH_PROP_ELEM, pinmux, Z_PINCTRL_STATE_PIN_INIT)};	\
> -
> +#define Z_PINCTRL_STATE_PINS_INIT(node_id, prop)                                                   \
> +	{DT_FOREACH_CHILD_VARGS(DT_PHANDLE(node_id, prop), DT_FOREACH_PROP_ELEM, pinmux,           \
> +				Z_PINCTRL_STATE_PIN_INIT)};
> ```
> 
> Are they valid.
> 
> Also, what is why is twister failing?
> 
> ```
>  Run tests with twister / twister-build (1) (pull_request_target) Failing after 39m 
> ```
> 
> Mostly this looks good to me. Once we clarify the issues above and have the HAL merged this is ready to go.

No twister failing on my local run.

Zhiqiang-Hou commented:
The twister failing should be caused by the hal_nxp revision is not including the depended patches.

dbaluta commented:
@Zhiqiang-Hou 

```
[Merge branch 'main' into develop/imx95-support](https://github.com/zephyrproject-rtos/zephyr/pull/74920/commits/6180c57ea786b9e5c706a6f9adf75c7e0af14d2e)

```

This commit shouldn't be present in your pull request.
```
Please run 
```
$ git fetch origin
git pull --rebase

```

If you want to rebase your patches on latest main branch

Zhiqiang-Hou commented:
It seems introdued by resolving the conflicts in west.yml through the 'Resolve conflicts' button. will rebase to remove that commit.

Zhiqiang-Hou commented:
The PR #77344 has fixed the twister failing issues. Thanks @LaurentiuM1234

LaurentiuM1234 commented:
https://github.com/zephyrproject-rtos/zephyr/pull/77344 fixes SCMI-related issues (sorry!), but it exposes some other issues with the SMP variant. See run from: https://github.com/zephyrproject-rtos/zephyr/pull/77343

JiafeiPan commented:
Hi, @yangbolu1991 , need to rebase this PR as https://github.com/zephyrproject-rtos/zephyr/pull/76829 has been merged, I mean need to reorg imx93 m33 code layout in soc directory.

yangbolu1991 committed:
soc: nxp: imx: create new directory for imx93

Created new directory for imx93 under imx9, as imx93
is one soc of imx9 series.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

Zhiqiang-Hou commented:
> #77344 fixes SCMI-related issues (sorry!), but it seems to expose some other issues with the SMP variant. See run from: #77343
Will have a look, thanks for the info!

yangbolu1991 committed:
soc: nxp: imx: clang-format imx93 code files

clang-format imx93 code files.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 commented:
Updated to v10. Changes,

- clang-format code files changed.
- Fixed smp board twister issue.
  - cherry-picked https://github.com/zephyrproject-rtos/zephyr/pull/77344 for CI. Will drop once it's merged. @LaurentiuM1234 
- Fixed pinctrl driver input register configuration.
- Rebased to address conflict.

Dependency
- hal patch: https://github.com/zephyrproject-rtos/hal_nxp/pull/406

Thank you very much.

LaurentiuM1234 commented:
https://github.com/zephyrproject-rtos/zephyr/pull/77344 got merged. Feel free to drop 15f0711eb5f8334e31c4b723126acd968326c383 and rebase.

dleach02 commented:
If this PR is dependent on a HAL PR then you need to represent that with an update to the west.yml and have it point to the PR by changing the revision to "revision: pull/406/head".

Zhiqiang-Hou commented:
Hi @dleach02 
Yes, I've already updated the west.yml to make hal_nxp point to "pull/406/head".

Zhiqiang-Hou commented:
v11 Changes:

- Fixed twister failing issue on tests/kernel/semaphore/sys_sem.

- Dropped the patch cherry-picked from https://github.com/zephyrproject-rtos/zephyr/pull/77344, since it has been merged.

- Rebased the PR.

Dependency:

hal patch: https://github.com/zephyrproject-rtos/hal_nxp/pull/406

Zhiqiang-Hou committed:
test: kernel/sem/sys_sem: custom MAX_THREAD_BYTES for imx95 SMP

Increased the value of CONFIG_MAX_THREAD_BYTES to avoid
triggering twister failing.

Signed-off-by: Hou Zhiqiang <Zhiqiang.Hou@nxp.com>

dleach02 commented:
Please address the CI issues

Zhiqiang-Hou committed:
test: kernel/sem/sem: custom MAX_THREAD_BYTES for imx95 SMP

Increased the value of CONFIG_MAX_THREAD_BYTES to avoid
triggering twister failing.

Signed-off-by: Hou Zhiqiang <Zhiqiang.Hou@nxp.com>

Zhiqiang-Hou commented:
v12 Changes:

- Fixed twister failing issue on tests/kernel/semaphore/semaphore.

- Rebased the PR.

Dependency:

- hal patch: https://github.com/zephyrproject-rtos/hal_nxp/pull/406

Zhiqiang-Hou commented:
v13 Changes:
- Fixed twister failing issue on tests/kernel/common.
- Rebased the PR.

Dependency:
- hal patch: https://github.com/zephyrproject-rtos/hal_nxp/pull/406

Zhiqiang-Hou committed:
test: kernel/common: custom MAX_THREAD_BYTES for imx95 SMP

Increased the value of CONFIG_MAX_THREAD_BYTES to avoid
triggering twister failing.

Signed-off-by: Hou Zhiqiang <Zhiqiang.Hou@nxp.com>

yangbolu1991 commented:
Updated to v14. Changes,
- Rebased to address conflict.

Dependency
- hal patch: https://github.com/zephyrproject-rtos/hal_nxp/pull/406

Thanks.

yangbolu1991 commented:
Updated to v15. Changes,
- Rebased to address conflict.

Dependency
- hal patch: https://github.com/zephyrproject-rtos/hal_nxp/pull/406

Thanks.

yangbolu1991 commented:
Any other comments on the pull request?
Thank you very much.

dbaluta commented:
Looking good and stable now. @dleach02  I think we are fine to move on and first merge the hal PR:

https://github.com/zephyrproject-rtos/hal_nxp/pull/406

JiafeiPan commented on boards/nxp/imx95_evk/imx95_evk_mimx9596_a55_smp.dts:
```diff
@@ -0,0 +1,37 @@
+	};
+
+	psci {
+		compatible = "arm,psci-0.2";
```
Sorry for late comments, it should be psci v1.1, please fix, thanks.

Zhiqiang-Hou replied:
will update the psci version. Thanks!

Zhiqiang-Hou commented:
Updated to v16:

- Updated PSCI version to v1.1.

Dependency:

- hal patch: https://github.com/zephyrproject-rtos/hal_nxp/pull/406

mmahadevan108 commented on drivers/pinctrl/Kconfig.imx:
Is there a binding file for this driver?

yangbolu1991 replied:
@mmahadevan108 
The new added pinctrl imx scmi driver is scmi based driver.
It depends on ARM_SCMI_PINCTRL_HELPERS, the binding file is,
dts/bindings/firmware/arm,scmi-pinctrl.yaml

Thanks a lot.

yangbolu1991 commented:
Updated to v17. Changes,
- Since hal patches were merged, updated west.yml

Thanks a lot.

mmahadevan108 commented:
Its unclear why the changes to the test config files are required. This would impact multiple platforms. 
@hakehuang can you please review and comment.

Zhiqiang-Hou commented:
> Its unclear why the changes to the test config files are required. This would impact multiple platforms. @hakehuang can you please review and comment.

These changes are to fix the twister failling issues, the imx95evk board default config has given a working CONFIG_MAX_THREAD_BYTES, but since the one in test prj.config overrides the board default config, we have to change the one in tests config.
As these changes just increase the size according to the twister suggestion, I think, it won't impact other platforms.

And can someone help us understand the relation between CONFIG_MAX_THREAD_BYTES and the enabled SMP cores?

hakehuang commented:
@Zhiqiang-Hou why there are imx93 code changes in this PR? Is there any inter-dependencies?

hakehuang commented on tests/kernel/common/prj.conf:
```diff
@@ -6,4 +6,4 @@ CONFIG_BOOT_DELAY=500
 CONFIG_TEST_USERSPACE=y
 CONFIG_BOUNDS_CHECK_BYPASS_MITIGATION=y
-CONFIG_MAX_THREAD_BYTES=3
+CONFIG_MAX_THREAD_BYTES=4
```
do not change the prj config, instead create a board folder and use board.conf to override it.

Zhiqiang-Hou replied:
Yes, will update it. Thanks for the good suggestion!

hakehuang commented on tests/kernel/semaphore/semaphore/prj.conf:
```diff
@@ -3,4 +3,4 @@ CONFIG_IRQ_OFFLOAD=y
 CONFIG_ZTEST_FATAL_HOOK=y
 CONFIG_PIPES=y
-CONFIG_MAX_THREAD_BYTES=3
+CONFIG_MAX_THREAD_BYTES=4
```
same as above

hakehuang commented on tests/kernel/semaphore/sys_sem/prj.conf:
```diff
@@ -1,4 +1,4 @@
 CONFIG_TEST_USERSPACE=y
 CONFIG_ZTEST_FATAL_HOOK=y
-CONFIG_MAX_THREAD_BYTES=3
+CONFIG_MAX_THREAD_BYTES=4
```
same as above

Zhiqiang-Hou commented:
> @Zhiqiang-Hou why there are imx93 code changes in this PR? Is there any inter-dependencies?

No functionality change on imx93, just reorganized the code to put the imx93 and imx95 under the imx9 family directory.

hakehuang commented:
> No functionality change on imx93, just reorganized the code to put the imx93 and imx95 under the imx9 family directory.

I would suggest to at least update the PR description

Zhiqiang-Hou commented:
V18:

- Addressed hakehuang's comments.
- Rebased.

dbaluta commented:
> And can someone help us understand the relation between CONFIG_MAX_THREAD_BYTES and the enabled SMP cores?

My opinion on this is that we should put those changes in a separate PR with a proper explanation! Otherwise, a lot of people will miss those changes.

This PR has already too many patches and start to be hard to follow and review.

dbaluta commented:
@yangbolu1991 I would suggest you split this in at least 3 PR:

- CI patches
- imx95 support + imx93 code reorg
- SMP support.

Zhiqiang-Hou commented:
> > No functionality change on imx93, just reorganized the code to put the imx93 and imx95 under the imx9 family directory.
> 
> I would suggest to at least update the PR description

sure

Zhiqiang-Hou commented:
> @yangbolu1991 I would suggest you split this in at least 3 PR:
> 
> * CI patches
> * imx95 support + imx93 code reorg
> * SMP support.

Do you think split into the following PRs is reasonable?
1. imx9/imx93 reorg
2. imx95 support
3. imx95 SMP support + CI patches (which are specific for imx95 SMP)

Thanks!

mmahadevan108 commented:
@hakehuang, can you revisit this PR to see if your comments are addressed. 
The test changes have been moved to a board level overlay file which is better than changing the base config file.
@Zhiqiang-Hou, can you address the CI failures.

yangbolu1991 commented:
> @yangbolu1991 I would suggest you split this in at least 3 PR:
> 
> * CI patches
> * imx95 support + imx93 code reorg
> * SMP support.

@dbaluta ,
Zhiqiang had made the CI patches imx95 specific. Now the PR code changes were only imx95 specific.
Is it okay to keep all patches? The changes had been review for many cycles.
Thanks.

Zhiqiang-Hou commented:
> @Zhiqiang-Hou, can you address the CI failures.
Sure

Zhiqiang-Hou commented:
V19:
- Shortened the subject of the patches triggered CI failure.
- Updated the revsion of hal_nxp in west.yml.
- Rebased.

dbaluta commented:
> @dbaluta ,
> Zhiqiang had made the CI patches imx95 specific. Now the PR code changes were only imx95 specific.
> Is it okay to keep all patches? The changes had been review for many cycles.
> Thanks.

Oh, you are right! Yes it is OK to keep the changes here.

Zhiqiang-Hou commented:
v20:

- Fixed CI failure caused by change log of the west.yml updating patch.

yangbolu1991 committed:
west.yml: hal_nxp: update to the latest

Updated hal_nxp to the latest:
30bdef6d s32: support LPI2C for s32z27x build with MCUX driver

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

dleach02 commented:
@hakehuang has @Zhiqiang-Hou addressed your change requests?

mmahadevan108 reviewed (changes requested):
Will block this to Synchronize the merge-order with the HAL

hakehuang commented:
> @hakehuang has @Zhiqiang-Hou addressed your change requests?

@dleach02 Yes.

mmahadevan108 commented:
> What do you mean by this. HAL PR was already merged and hash commit was updated in this PR to reflect the HAL changes.
> 
> Are there any other PRs that update the HAL hash so that we need to sync with those?
> 
> Later edit: I think it is about this:

Waiting for this PR to merge: https://github.com/zephyrproject-rtos/zephyr/pull/77218

mmahadevan108 commented:
Kindly rebase to fix the `west.yml` conflict

mmahadevan108 reviewed (dismissed):
Please rebase

yangbolu1991 commented:
Updated to v21. Changes,
- Rebased to resolve conflict.

Thanks.

nordicjm commented on boards/nxp/imx95_evk/board.cmake:
```diff
@@ -0,0 +1 @@
+# SPDX-License-Identifier: Apache-2.0
```
get rid of this file

nordicjm commented on boards/nxp/imx95_evk/CMakeLists.txt:
```diff
@@ -0,0 +1 @@
+# SPDX-License-Identifier: Apache-2.0
```
get rid of this file

nordicjm commented on boards/nxp/imx95_evk/doc/index.rst:
```diff
@@ -0,0 +1,262 @@
+Supported Features
+==================
+
+The Zephyr imx95_evk_m7 board configuration supports the following hardware features:
```
this is not the board name at all, put \`\` around it, put the actual board name you supply to west and remove board configuration because such a term has never existed in zephyr and replace with board target

nordicjm commented on boards/nxp/imx95_evk/imx95_evk_mimx9596_a55_defconfig:
```diff
@@ -0,0 +1,36 @@
+CONFIG_UART_CONSOLE=y
+
+CONFIG_CLOCK_CONTROL=y
+CONFIG_PINCTRL=y
```
```suggestion
```
`CONFIG_PINCTRL=y` should not be present, remove it

nordicjm commented on boards/nxp/imx95_evk/imx95_evk_mimx9596_a55_smp_defconfig:
```diff
@@ -0,0 +1,43 @@
+CONFIG_UART_CONSOLE=y
+
+CONFIG_CLOCK_CONTROL=y
+CONFIG_PINCTRL=y
```
as above and to all board files

nordicjm reviewed (commented):
Seemingly I missed this PR and it should not have been merged, please fix the issues above in a follow up PR

yangbolu1991 commented:
> Seemingly I missed this PR and it should not have been merged, please fix the issues above in a follow up PR

Thanks @nordicjm for the comments.
I created one more PR https://github.com/zephyrproject-rtos/zephyr/pull/78276

Going through the zephyr boards, it looks like these are common problem.
One more question, why should remove CONFIG_PINCTRL=y ?
Or it's not the right place to put it?

Thank you very much.

nordicjm commented:
> One more question, why should remove CONFIG_PINCTRL=y ?

Drivers should select pinctrl, it is not a board that needs it, is it drivers that need it, if I can go in and de-select the pinctrl Kconfig and it breaks the build then the drivers are wrong.

yangbolu1991 commented:
> > One more question, why should remove CONFIG_PINCTRL=y ?
> 
> Drivers should select pinctrl, it is not a board that needs it, is it drivers that need it, if I can go in and de-select the pinctrl Kconfig and it breaks the build then the drivers are wrong.

Thanks for clarifying. It makes sense.
I will add on more patch for this.
I remembered for linux/u-boot, there is "make savedefconfig" to refresh. If only zephyr has that :)
